src/api/scenarioData/scenarioData.service.ts

```diff
@@ -0,0 +1,379 @@
+    try {
+      const scenarioData = await this.scenarioDataModel.findOne({ _id: scenarioId }).exec();
+      if (!scenarioData) {
+        throw new Error('Scenario data not found');
+      }
+      return scenarioData;
+    } catch (error) {
+      throw new Error(`Error getting scenario data: ${error.message}`);
+    }
+  }
+
+  async createScenarioData(scenarioData: ScenarioData): Promise<ScenarioData> {
+    try {
+      const newScenarioData = new this.scenarioDataModel(scenarioData);
+      const response = await newScenarioData.save();
+      return response;
+    } catch (error) {
+      console.error('Error in createScenarioData:', error);
+      throw new Error(`Failed to add scenario data: ${error.message}`);
+    }
+  }
+
+  async updateScenarioData(scenarioId: string, updatedData: Partial<ScenarioData>): Promise<ScenarioData> {
+    try {
+      const existingScenarioData = await this.scenarioDataModel.findOne({ _id: scenarioId }).exec();
+      if (!existingScenarioData) {
+        throw new Error('Scenario data not found');
+      }
+      Object.assign(existingScenarioData, updatedData);
+      return await existingScenarioData.save();
+    } catch (error) {
+      throw new Error(`Failed to update scenario data: ${error.message}`);
+    }
+  }
+
+  async deleteScenarioData(scenarioId: string): Promise<void> {
+    try {
+      const objectId = new Types.ObjectId(scenarioId);
+      const deletedScenarioData = await this.scenarioDataModel.findOneAndDelete({ _id: objectId });
+      if (!deletedScenarioData) {
+        throw new Error('Scenario data not found');
+      }
+      return;
+    } catch (error) {
+      throw new Error(`Failed to delete scenario data: ${error.message}`);
+    }
+  }
+
+  async getScenariosByRuleId(ruleId: string): Promise<ScenarioData[]> {
+    try {
+      return await this.scenarioDataModel.find({ ruleID: ruleId }).exec();
+    } catch (error) {
+      throw new Error(`Error getting scenarios by rule ID: ${error.message}`);
+    }
+  }
+
+  async getScenariosByFilename(goRulesJSONFilename: string): Promise<ScenarioData[]> {
+    try {
+      return await this.scenarioDataModel.find({ goRulesJSONFilename: goRulesJSONFilename }).exec();
+    } catch (error) {
+      throw new Error(`Error getting scenarios by filename: ${error.message}`);
+    }
+  }
+
+  //handle special characters that may be entered by end users
+  replaceSpecialCharacters(input: string, replacement: string): string {
+    const specialChars = /[\n\r\t\f,]/g;
+    return input.replace(specialChars, (match) => {
+      if (match === ',') {
+        return '-';
+      }
+      return replacement;
+    });
+  }
+
+  /**
+   * Runs decisions for multiple scenarios based on the provided rules JSON file.
+   * Retrieves scenarios, retrieves rule schema, and executes decisions for each scenario.
+   * Maps inputs and outputs from decision traces to structured results.
+   */
+  async runDecisionsForScenarios(
```

Comment:
I think you should spend a bit of time to clean up this function because it's a bit longer and harder to read than it needs to be. Put the helper util functions somewhere else and make sure no repetition is happening. Then this function can just read something like:

```
async function processScenarios(scenarios: Scenario[]) {
  const results: Record<string, any> = {};

  for (const scenario of scenarios) {
    const formattedVariablesObject = reduceToCleanObj(scenario.variables);
    const formattedExpectedResultsObject = reduceToCleanObj(scenario.expectedResults);

    try {
      const decisionResult = await this.decisionsService.runDecisionByFile(
        scenario.goRulesJSONFilename,
        formattedVariablesObject,
        { trace: true },
      );

      const resultMatches = isEqual(decisionResult.result, formattedExpectedResultsObject) || Object.keys(expectedResultsObject).length === 0;

      const scenarioResult = {
        inputs: mapTraces(decisionResult.trace, 'input'),
        outputs: mapTraces(decisionResult.trace, 'output'),
        expectedResults: formattedExpectedResultsObject || {},
        result: decisionResult.result || {},
        resultMatch: resultMatches,
      };

      results[scenario.title] = scenarioResult;
    } catch (error) {
      console.error(`Error running decision for scenario ${scenario._id}: ${error.message}`);
      results[scenario._id] = { error: error.message };
    }
  }

  return results;
}
```

You might want to do something similar with the other functions further down in this file.